### DISCUSSION TITLE
Why not await Animation.RunAsync when IterationCount is Infinite

### DISCUSSION THREAD

**Discussion by lindexi** [AvaloniaUI/Avalonia#16757] [General]
As you can see, the `Animation.RunAsync` will set the excpetion to TaskCompletionSource when the IterationCount is Infinite

https://github.com/AvaloniaUI/Avalonia/blob/2d516027602698b1aa817f46edc62fd9d337c66a/src/Avalonia.Base/Animation/Animation.cs#L342-L343

I have other ideas, we can await the Animation.RunAsync when IterationCount is Infinite. And we can finish the RunAsync by cancellationToken be set.

By the ways, the other issues is when I receive the InvalidOperationException with `Looping animations must not use the Run method.`, that I do not know what should I do.

**Comment by h0lg** [comment]
Similar question here: How do I **start and stop** an animation with `IterationCount.Infinite`, e.g. to animate an active remote search input?
I can get something to work by setting `IterationCount` to 0 intially, then to `IterationCount.Infinite` and use `RunAsync` with a cancellation token to start the animation, ignore the returned task and use the token to stop the animation. Also, resetting `IterationCount` to 0 seems to stop the animation.
The `Looping animations must not use the Run method.` error message still trips me up tho.

And even if this seems to work most of the time, there seems to be no way to observe (at code runtime) **if the animation is currently running**. That would be helpful to decide whether to let it run or restart such a longer living animation.

**Reply by maxkatz6** [reply]
> How do I start and stop an animation with IterationCount.Infinite

Bind animation to a specific style selector, with a class-name or property activator for example.
And then toggle this class-name/property. It is how it's normally done with Avalonia XAML animations. You don't need to create and manually run animations in C#. And in case of infinite animations - you can't (unless you add it to styles in the same way as in XAML).

**Reply by lindexi** [reply]
@maxkatz6 Some logic request to control the animation start in the C# code.

**Reply by h0lg** [reply]
@maxkatz6 Thanks for your feedback! I'd like to avoid XAML in my project, but may try it for this corner case. I have however successfully created a [style with a class selector](https://docs.avaloniaui.net/docs/reference/styles/style-selector-syntax#by-style-class) and added the code-created (F# actually since I'm using it via [Fabulous Avalonia](https://github.com/fabulous-dev/Fabulous.Avalonia)) animation to it via `Style.Animations`. The animation starts if I add the trigger class - but doesn't stop reliably when removing it.

On another note, I've had another idea for **working around** the animation not stopping in my scenario, which is an `AutoCompleteBox` that I want to animate while its `AsyncPopulator` is running. I could periodically trigger the animation (with an `IterationCount` other than `Infinite`) manually while the `AsyncPopulator` is running.

**Reply by h0lg** [reply]
> Some logic request to control the animation start in the C# code.

@lindexi That is hard to understand. Is that a question or a statement? Maybe try rephrasing.

**Reply by lindexi** [reply]
Some business logic requires manual control over the start and stop of animations. For instance, in applications with visually stunning interfaces, an animation might start playing when the user clicks button A and only stop when the user clicks button B. Since developers cannot predict how long it will take for the user to click button B, they need the animation to play indefinitely. The application code will then handle the interaction with the buttons to determine whether the animation should stop.

**Reply by leonidcan** [reply]
And infinite animations create memory leaks when defined in AXAML; this is a very old bug. Will it ever be fixed?

**Comment by kkwpsv** [comment]
This code will trigger TaskScheduler.UnobservedTaskException and increase the pressure on finalizers.

**Comment by stevemonaco** [comment]
For infinite animations, consider using compositor-based animations with `AnimationIterationBehavior.Forever` instead.

**Reply by lindexi** [reply]
@stevemonaco The business code cannot manually stop the ompositor-based animation. Please consider such a situation: the UI interface needs to play a certain animation until the user clicks a certain button to stop.

**Reply by maxkatz6** [reply]
I think it would be a reasonable feature request to add an extra check for `cancellationToken.CanBeCancelled`.
If token can be cancelled - allow IterationCount.Infinite

**Reply by stevemonaco** [reply]
I fail to see how that's a problem. The business code (or the UI code calling it) should produce a side effect that the UI will observe. It's up to the dev to bridge that change so that the UI is aware.

For example, I have idle animations via compositor: `<LoadingAnimation IsActive="{Binding IsLoaded}" />`

You can subscribe: `IsActiveProperty.Changed.AddClassHandler<LoadingAnimation>(static (animation, args) => IsActiveChanged(animation, args));`

Then the handler applies or removes the composition animation.
